azure-spring-boot/src/main/java/com/microsoft/azure/keyvault/spring/KeyVaultOperation.java

```diff
@@ -3,91 +3,148 @@
-import org.springframework.util.Assert;
-import org.springframework.util.StringUtils;
 
-import java.util.ArrayList;
-import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Locale;
-import java.util.Optional;
-import java.util.stream.Collectors;
+import java.util.Timer;
+import java.util.TimerTask;
 import java.util.stream.Stream;
 
 @Slf4j
 public class KeyVaultOperation {
-    
+
     /**
      * Stores the case sensitive flag.
      */
     private final boolean caseSensitive;
 
-    private final SecretClient keyVaultClient;
-    private final String vaultUri;
-    private volatile List<String> secretNames;
-    private final boolean secretNamesAlreadyConfigured;
-    private final long secretNamesRefreshIntervalInMs;
-    private volatile long secretNamesLastUpdateTime;
+    /**
+     * Stores the properties.
+     */
+    private LinkedHashMap<String, String> properties = new LinkedHashMap<>();
+
+    /**
+     * Stores the secret client.
+     */
+    private final SecretClient secretClient;
+
+    /**
+     * Stores the secret keys.
+     */
+    private List<String> secretKeys;
 
+    /**
+     * Constructor.
+     *
+     * @param secretClient the Key Vault secret client.
+     * @param refreshInMillis the refresh in milliseconds (0 or less disables
+     * refresh).
+     * @param caseSensitive the case sensitive flag.
+     */
     public KeyVaultOperation(
-            final SecretClient keyVaultClient,
-            String vaultUri,
-            final long secretKeysRefreshIntervalInMs,
-            final List<String> secretNames,
-            boolean caseSensitive
-    ) {
+            final SecretClient secretClient,
+            final long refreshInMillis,
+            List<String> secretKeys,
+            boolean caseSensitive) {
+
         this.caseSensitive = caseSensitive;
-        this.keyVaultClient = keyVaultClient;
-        // TODO(pan): need to validate why last '/' need to be truncated.
-        this.vaultUri = StringUtils.trimTrailingCharacter(vaultUri.trim(), '/');
-        this.secretNames = Optional.ofNullable(secretNames)
-                .map(Collection::stream)
-                .orElseGet(Stream::empty)
-                .map(this::toKeyVaultSecretName)
-                .distinct()
-                .collect(Collectors.toList());
-        this.secretNamesAlreadyConfigured = !this.secretNames.isEmpty();
-        this.secretNamesRefreshIntervalInMs = secretKeysRefreshIntervalInMs;
-        this.secretNamesLastUpdateTime = 0;
+        this.secretClient = secretClient;
+        this.secretKeys = secretKeys;
+
+        if (refreshInMillis > 0) {
+            final Timer timer = new Timer();
+            final TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    refreshProperties();
+                }
+            };
+            timer.scheduleAtFixedRate(task, 0, refreshInMillis);
```

Comment:
If `refreshInMillis` equal to 1 second, and `task` will cost 1 minute to finish, then many task will run at the same time? That's not good.

Reply:
And it's not necessary to `refreshProperties()` again and again if the app only need to read the properties once.

(cc @jialindai FYI.)

Reply:
I mean, if `getPropertyNames()` and `getProperty` are not called, `refreshProperties()` should not execute.

Reply:
> I disagree. A refresh interval is exactly that, aka the interval after which the keys/values are refreshed from the backend. So not going to change this. If the application developer does not need a refresh of the keys/values after startup the refresh interval can be set to 0.

If user configure refresh time to 5 minutes, and getPropertyNames() and getProperty are not called, refreshProperties() every 5 minutes is waste of resource, and will be a burden to key-vault service.

Maybe `cache-time` is better configure item than `refresh-interval`?

Reply:
>From the JavaDoc of Timer, "Corresponding to each Timer object is a single background thread that is used to execute all of the timer's tasks, sequentially."

Then what happen if refresh-interval  = 1 minute and refresh operation need 2 minutes?

Reply:
I feel you are looking to implement a TTL for each secret we would have to rearchitect this. Can we proceed as it currently stands? And IF we decide that TTL is appropriate to do that later?

Reply:
If the refresh takes longer than 2 minutes I would argue we have a problem with Azure KeyVault. But even with that said why on earth would anyone want to refresh the cache every minute? If there are doing that then they are using this integration for something other than its intended purpose and they would be better off using the SDK to interact with Azure KeyVault.

Reply:
> I feel you are looking to implement a TTL for each secret, if so we would have to rearchitect this.

Not for each secret, use current configure item `refresh-interval` as TTL of all secrets is my purpose.

> Can we proceed as it currently stands? And IF we decide that TTL is appropriate to do that later?

Can we use  current configure item `refresh-interval` as TTL of all secrets?

Reply:
> But even with that said why on earth would anyone want to refresh the cache every minute?

Just as I said above: our customer may update key-value pair in key-vault (manually or by java code), then get the pair by environment.getProperty() minutes latter. if customer can not get the updated data, it may make our customer confused.

Reply:
> > I feel you are looking to implement a TTL for each secret, if so we would have to rearchitect this.
> 
> Not for each secret, use current configure item `refresh-interval` as TTL of all secrets is my purpose.
> 
> > Can we proceed as it currently stands? And IF we decide that TTL is appropriate to do that later?
> 
> Can we use current configure item `refresh-interval` as TTL of all secrets?

Great then I can leave the code I wrote as is as that is what the refresh interval does. It refreshes all the secrets after a specific time (Time To Live)

Reply:
> > But even with that said why on earth would anyone want to refresh the cache every minute?
> 
> Just as I said above: our customer may update key-value pair in key-vault (manually or by java code), then get the pair by environment.getProperty() minutes latter. if customer can not get the updated data, it may make our customer confused.

See my comments above

Reply:
>  It refreshes all the secrets after a specific time (Time To Live)

To save the resource of network and key-vault server:
> I mean, if getPropertyNames() and getProperty are not called, refreshProperties() should not execute.